include/deal.II/lac/solver_fire.h

```diff
@@ -0,0 +1,392 @@
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/*!@addtogroup Solvers */
+/*@{*/
+
+/**
+ * FIRE (Fast Inertial Relaxation Engine) for minimization of (potentially
+ * non-linear) objective function $E(U)$, $U$ is a vector of $n$ variables
+ * ($n$ is the number of variables of the objective function).
+ * Like all other solver classes, it can work on any kind of vector and matrix
+ * as long as they satisfy certain requirements (for the requirements on
+ * matrices and vectors in order to work with this class, see the documentation
+ * of the Solver base class). The type of the solution vector must be passed as
+ * template argument, and defaults to dealii::Vector<double>.
+ *
+ * FIRE is a damped dynamics method described in
+ * <a href="https://doi.org/10.1103/PhysRevLett.97.170201">Structural
+ * Relaxation Made Simple</a> by Bitzek et al. 2006, typically used to find
+ * stable equilibrium configurations of atomistic systems in computational
+ * material science. Starting from a given initial configuration of the
+ * atomistic system, the algorithm relies on inertia to obtain (nearest)
+ * configuration with least potential energy.
+ *
+ * The problem of solving the system of linear equations $AU = F$ for $U$
+ * can be re-casted into the problem of finding $U$ for which the quadratic
+ * function $\frac{1}{2} U^{T}AU - U^{T}F$ is minimized.
+ *
+ * Notation:
+ * The global vector of unknowns:   $U$.        <BR>
+ * Objective function:              $E(U)$.     <BR>
+ * Rate of change of unknowns:      $V$.        <BR>
+ * Gradient of the objective
+ * function w.r.t unknowns:         $G$.        <BR>
+ * Mass matrix:                     $M$.        <BR>
+ * Initial guess of unknowns:       $U_0$.      <BR>
+ * Time step:                       $\Delta t$. <BR>
+ *
+ * Given initial values for $\Delta t$, $\alpha = \alpha_0$, $U = U_0$ and $V=0$
+ * along with a given mass matrix $M$,
+ * FIRE algorithm is as follows,
+ * 1. Calculate $G = \nabla E(U)$ and check for convergence.
+ *    Calculate $U$ and $V$ using any common Molecular Dynamics time integrator.
+ *    Here we use simple Euler integration step, <BR>
+ *        $U = U + \Delta t V$,                  <BR>
+ *        $V = V + \Delta t M^{-1} \cdot G$.
+ * 2. Calculate $P = G \cdot V$.
+ * 3. Set $V = (1-\alpha) V + \alpha \frac{|V|}{|G|} G$.
+ * 4. If $P \leq 0$ and number of steps since P was non-negative is larger than
+ *    certain value, then increase time step $\Delta t$ and decrease $\alpha$.
+ * 5. If $P>0$, then decrease the time step, freeze the system i.e., V = 0 and
+ *    reset $\alpha = \alpha_0$.
+ * 6. Return to 1.
+ *
+ * Alse see
+ * <a href="http://onlinelibrary.wiley.com/doi/10.1002/pamm.201110246/full">
+ * Energy-Minimization in Atomic-to-Continuum Scale-Bridging Methods </a> by
+ * Eidel et al. 2011.
+ *
+ * @author Vishal Boddu 2017
+ */
+template<typename VectorType = Vector<double> >
+class SolverFIRE : public Solver<VectorType>
+{
+
+public:
+
+  struct AdditionalData
```

Comment:
Can you document this class and its constructor? It doesn't require much, just a sentence or two each.